### PR TITLE
Fix more context-related generics

### PR DESCRIPTION
**Pull request by glasser**
- GraphQLExecutor took a default context argument and thus wasn't always
  properly assigned the right context. Fixing this ended up adding
  `<TContext>` to GatewayInterface and thus SchemaManager and ServerState,
  etc.

- A few things had `<TContext>` without `extends BaseContext`.

- It's reasonably likely that we are going to be changing ApolloServer
  and ApolloServerPlugin to be `in out` (invariant) in their context
  type, so to ready for that, change the built-in plugins to return
  `<TContext extends BaseContext>` instead of just `<BaseContext>`.

### CODE REVIEW

File: packages/server/src/types.ts [hunk 1]
```diff
@@ -45,18 +45,21 @@ export type AnyFunctionMap = {
...
 };
 
 export type SchemaLoadOrUpdateCallback = (schemaContext: {
   apiSchema: GraphQLSchema;
   coreSupergraphSdl: string;
 }) => void;
 
-export interface GatewayInterface {
-  load(options: { apollo: ApolloConfig }): Promise<GraphQLServiceConfig>;
+export interface GatewayInterface<TContext extends BaseContext> {
```
Comment (trevor-scheer) [inline packages/server/src/types.ts:59]: This change forces an update on gateway's usage of the interface, which is maybe ok? I know we've removed some back-compat with older gateway versions. I'm wondering if there's value in a default `= BaseContext` here, or if that wouldn't work anyway once we're using `in out`.

**Review by trevor-scheer — COMMENTED** [review]
Awesome!

File: packages/server/src/types.ts (hunk 1 above)
Comment (glasser) [reply packages/server/src/types.ts:59]: I kinda assume there's going to be a minimum Gateway version required for AS4 anyway?

**Comment by glasser** [comment]
Feel free to rebase and merge if you're doing anything that might overlap with this today. (Conflicts vs the other PR I merged should be trivial)

File: packages/server/src/types.ts (hunk 1 above)
Comment (trevor-scheer) [reply packages/server/src/types.ts:59]: Yeah, I think that's a fine answer to this question. Very reasonable that moving to AS4 requires you be on a latest version of Gateway.

**Pull request merged**